File: test_preferences.py
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
from solders.keypair import Keypair
from solders.pubkey import Pubkey
from solders.system_program import ID as SYS_PROGRAM_ID
from anchorpy import Provider, Wallet, Program
import json
import asyncio

# Load the IDL
async def load_idl():
    with open("programs/userpreferences/target/idl/user_preferences.json", "r") as f:
        raw_idl = json.load(f)
        
        class Argument:
            def __init__(self, arg_dict):
                self.name = arg_dict.get('name')
                rust_type = arg_dict.get('type')
                if isinstance(rust_type, dict) and 'vec' in rust_type:
                    # Handle vector types (for whitelisted_addresses and whitelisted_currencies)
                    if rust_type['vec'] == 'pubkey':
                        self.ty = {'vec': 'publicKey'}
                    elif rust_type['vec'] == 'string':
                        self.ty = {'vec': 'string'}
                elif rust_type == 'u64':
                    self.ty = 'u64'
                else:
                    self.ty = rust_type

        class Account:
            def __init__(self, acc_dict):
                self.name = acc_dict.get('name')
                self.isMut = acc_dict.get('writable', False)
                self.isSigner = acc_dict.get('signer', False)
                if 'pda' in acc_dict:
                    self.pda = acc_dict['pda']

        class Instruction:
            def __init__(self, ix_dict):
                self.name = ix_dict.get('name')
                self.accounts = [Account(acc) for acc in ix_dict.get('accounts', [])]
                self.args = [Argument(arg) for arg in ix_dict.get('args', [])]
                self.discriminator = ix_dict.get('discriminator')

        class IDL:
            def __init__(self, raw):
                self.version = raw.get('metadata', {}).get('version')
                self.name = raw.get('metadata', {}).get('name')
                self.instructions = [Instruction(ix) for ix in raw.get('instructions', [])]
                self.accounts = raw.get('accounts', [])
                self.types = raw.get('types', [])
                self.metadata = raw.get('metadata', {})
                self.address = raw.get('address')

        return IDL(raw_idl)

async def main():
    # Initialize the client
    client = AsyncClient("http://localhost:8899", commitment=Commitment("confirmed"))
    
    # Load or create a keypair for testing
    keypair_path = "test-keypair.json"
    try:
        with open(keypair_path, "r") as f:
            content = f.read()
            if not content.strip():
                raise FileNotFoundError
            keypair_bytes = bytes(json.loads(content))
            keypair = Keypair.from_bytes(keypair_bytes)
    except (FileNotFoundError, json.JSONDecodeError):
        print("Creating new keypair...")
        keypair = Keypair()
        with open(keypair_path, "w") as f:
            json.dump(list(bytes(keypair)), f)

    # Create provider
    provider = Provider(client, Wallet(keypair))
    
    # Load the IDL and initialize the program
    idl = await load_idl()
    program_id = Pubkey.from_string("7XDgqx21jrGZSDbm4zvHzvnCti4NmenGdeHTbAJucx7R")
    program = Program(idl, program_id, provider)

    # Get the PDA for user preferences (using the user's public key)
    user_preferences_pda, bump = Pubkey.find_program_address(
        [b"user-preferences", bytes(keypair.pubkey)],
        program_id
    )

    print("\nCreating and updating User Preferences using a single account...")
    try:
        # Initialize the user preferences account
        await program.methods.initializeUserPreferences() \
            .accounts({
                "user_preferences": user_preferences_pda,
                "user": keypair.pubkey,
                "system_program": SYS_PROGRAM_ID,
            }) \
            .signers([keypair]) \
            .rpc()

        # Update the preferences with conservative values
        await program.methods.updatePreferences(
            1000,               # max_currency: 1000 units
            5,                  # max_transactions_per_day: 5
            [],                 # whitelisted_addresses: empty list
            ["USDC", "SOL"]     # whitelisted_currencies
        ).accounts({
            "user_preferences": user_preferences_pda,
            "user": keypair.pubkey,
        }).signers([keypair]).rpc()
        print("User preferences set successfully with one account!")
    except Exception as e:
        print(f"Error setting user preferences: {str(e)}")

    # Fetch and display the final state for the account
    try:
        account = await program.account["UserPreferences"].fetch(user_preferences_pda)
        print("\nFinal User Preferences State:")
        print(f"Max Currency: {account.max_currency}")
        print(f"Max Transactions Per Day: {account.max_transactions_per_day}")
        print(f"Whitelisted Currencies: {account.whitelisted_currencies}")
    except Exception as e:
        print(f"Error fetching final state: {str(e)}")

    await client.close()

if __name__ == "__main__":
    asyncio.run(main())

File: test_preferences.rs
use anchor_client::{
    solana_sdk::{
        commitment_config::CommitmentConfig,
        pubkey::Pubkey,
        signature::{Keypair, read_keypair_file, write_keypair_file},
        system_program,
    },
    Client, Program,
};
use std::{rc::Rc, str::FromStr};
use anyhow::Result;

// Your program's account struct
#[derive(Debug)]
pub struct UserPreferences {
    pub user: Pubkey,
    pub max_currency: u64,
    pub max_transactions_per_day: u64,
    pub whitelisted_addresses: Vec<Pubkey>,
    pub whitelisted_currencies: Vec<String>,
    pub bump: u8,
}

fn main() -> Result<()> {
    // Setup client
    let keypair = match read_keypair_file("test-keypair.json") {
        Ok(kp) => kp,
        Err(_) => {
            let new_keypair = Keypair::new();
            write_keypair_file(&new_keypair, "test-keypair.json")?;
            new_keypair
        }
    };

    // Initialize client
    let client = Client::new_with_options(
        "http://localhost:8899".to_string(),
        Rc::new(keypair),
        CommitmentConfig::confirmed(),
    );

    // Your program ID
    let program_id = Pubkey::from_str("7XDgqx21jrGZSDbm4zvHzvnCti4NmenGdeHTbAJucx7R")?;
    let program = client.program(program_id);

    // Find PDA
    let seeds = [
        b"user-preferences",
        program.payer().pubkey().as_ref(),
    ];
    let (pda, bump) = Pubkey::find_program_address(&seeds, &program_id);

    // Test Case 1: Conservative User
    println!("\nCreating Conservative User Preferences...");
    match initialize_preferences(&program, &pda) {
        Ok(_) => {
            match update_preferences(
                &program,
                &pda,
                1000,
                5,
                vec![],
                vec!["USDC".to_string(), "SOL".to_string()],
            ) {
                Ok(_) => println!("Conservative preferences set successfully!"),
                Err(e) => println!("Error setting conservative preferences: {}", e),
            }
        }
        Err(e) => println!("Error initializing preferences: {}", e),
    }

    // Test Case 2: Moderate User
    println!("\nUpdating to Moderate User Preferences...");
    match update_preferences(
        &program,
        &pda,
        5000,
        20,
        vec![],
        vec!["USDC".to_string(), "SOL".to_string(), "ETH".to_string(), "BTC".to_string()],
    ) {
        Ok(_) => println!("Moderate preferences set successfully!"),
        Err(e) => println!("Error setting moderate preferences: {}", e),
    }

    // Test Case 3: Aggressive User
    println!("\nUpdating to Aggressive User Preferences...");
    match update_preferences(
        &program,
        &pda,
        20000,
        50,
        vec![],
        vec!["USDC".to_string(), "SOL".to_string(), "ETH".to_string(), "BTC".to_string()],
    ) {
        Ok(_) => println!("Aggressive preferences set successfully!"),
        Err(e) => println!("Error setting aggressive preferences: {}", e),
    }

    // Fetch and display final state
    match program.account::<UserPreferences>(pda) {
        Ok(account) => {
            println!("\nFinal User Preferences State:");
            println!("Max Currency: {}", account.max_currency);
            println!("Max Transactions Per Day: {}", account.max_transactions_per_day);
            println!("Whitelisted Currencies: {:?}", account.whitelisted_currencies);
        }
        Err(e) => println!("Error fetching final state: {}", e),
    }

    Ok(())
}

fn initialize_preferences(program: &Program, pda: &Pubkey) -> Result<()> {
    program
        .request()
        .accounts(initialize_accounts(program, pda))
        .args(initialize_user_preferences_ix())
        .send()?;
    Ok(())
}

fn update_preferences(
    program: &Program,
    pda: &Pubkey,
    max_currency: u64,
    max_transactions_per_day: u64,
    whitelisted_addresses: Vec<Pubkey>,
    whitelisted_currencies: Vec<String>,
) -> Result<()> {
    program
        .request()
        .accounts(update_accounts(program, pda))
        .args(update_preferences_ix(
            max_currency,
            max_transactions_per_day,
            whitelisted_addresses,
            whitelisted_currencies,
        ))
        .send()?;
    Ok(())
}

// Helper functions to create instruction data
fn initialize_user_preferences_ix() -> anchor_client::anchor_lang::InstructionData {
    anchor_client::anchor_lang::InstructionData {
        program_id: anchor_client::anchor_lang::Id::new(0),
        data: vec![240, 126, 152, 62, 184, 25, 62, 149], // Your initialize discriminator
    }
}

fn update_preferences_ix(
    max_currency: u64,
    max_transactions_per_day: u64,
    whitelisted_addresses: Vec<Pubkey>,
    whitelisted_currencies: Vec<String>,
) -> anchor_client::anchor_lang::InstructionData {
    // Your update preferences instruction data
    anchor_client::anchor_lang::InstructionData {
        program_id: anchor_client::anchor_lang::Id::new(0),
        data: vec![16, 64, 128, 133, 19, 206, 101, 159], // Your update discriminator
    }
}

// Helper functions to create account metas
fn initialize_accounts<'a>(
    program: &Program,
    pda: &'a Pubkey,
) -> anchor_client::anchor_lang::ToAccountMetas {
    vec![
        AccountMeta::new(*pda, false),
        AccountMeta::new(program.payer().pubkey(), true),
        AccountMeta::new_readonly(system_program::ID, false),
    ]
}

fn update_accounts<'a>(
    program: &Program,
    pda: &'a Pubkey,
) -> anchor_client::anchor_lang::ToAccountMetas {
    vec![
        AccountMeta::new(*pda, false),
        AccountMeta::new(program.payer().pubkey(), true),
    ]
}